src/main/java/org/pesmypetcare/webservice/dao/usermanager/UserDaoImpl.java

```diff
@@ -0,0 +1,499 @@
+     *
+     * @param uid The unique identifier of the user
+     * @param batch The batch of writes
+     * @throws FirebaseAuthException If an error occurs when retrieving the data
+     */
+    private void deleteOldUsername(String uid, WriteBatch batch) throws FirebaseAuthException {
+        String oldUsername = myAuth.getUser(uid).getDisplayName();
+        batch.delete(usedUsernames.document(oldUsername));
+    }
+
+    /**
+     * Updates the display name.
+     *
+     * @param uid The unique identifier of the user
+     * @param newUsername The new username for the account
+     * @throws FirebaseAuthException If an error occurs when retrieving the data
+     */
+    private void updateDisplayName(String uid, String newUsername) throws FirebaseAuthException {
+        UserRecord.UpdateRequest updateRequest = getUserRecord(uid);
+        updateRequest.setDisplayName(newUsername);
+        myAuth.updateUserAsync(updateRequest);
+    }
+
+    /**
+     * Gets the update request for the user.
+     *
+     * @param uid The unique identifier of the user
+     * @return An update request for the user data
+     * @throws FirebaseAuthException If an error occurs when retrieving the data
+     */
+    private UserRecord.UpdateRequest getUserRecord(String uid) throws FirebaseAuthException {
+        return myAuth.getUser(uid).updateRequest();
+    }
+
+    /**
+     * Throws DatabaseAccessException if the user does not exist.
+     *
+     * @param userDoc The document snapshot of the user
+     * @throws DatabaseAccessException If the user does not exist
+     */
+    private void throwExceptionIfUserDoesNotExist(DocumentSnapshot userDoc) throws DatabaseAccessException {
+        if (!userDoc.exists()) {
+            throw new DatabaseAccessException(INVALID_USER, USER_DOES_NOT_EXIST_MESSAGE);
+        }
+    }
+
+    /**
+     * Updates the FCM token in all the groups the user is subscribed to.
+     * @param token The new FCM token
+     * @param currentToken The current FCM token
+     * @param batch The batch of writes
+     * @throws DatabaseAccessException When the operation is interrupted
+     * @throws DocumentException When the update fails
+     */
+    private void updateTokenInSubscribedGroups(String token, String currentToken, WriteBatch batch)
+        throws DatabaseAccessException, DocumentException {
+        ApiFuture<QuerySnapshot> subscribedGroups = collectionAdapter
+            .getDocumentsWhereArrayContains(Path.ofCollection(Collections.groups), "notification-tokens", currentToken);
+        try {
+            for (QueryDocumentSnapshot group : subscribedGroups.get().getDocuments()) {
+                batch.update(group.getReference(), "FCM", FieldValue.arrayRemove(currentToken), "FCM",
+                    FieldValue.arrayUnion(token));
+            }
+        } catch (InterruptedException e) {
+            throw new DatabaseAccessException("write-failed", e.getMessage());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The String "write-failed" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5561439)